cuda/test/factorization/par_ilu_kernels.cpp

```diff
@@ -0,0 +1,239 @@
+        auto coo_cuda = Coo::create(cuda);
+        csr_cuda->convert_to(gko::lend(coo_cuda));
+        initialize_lu(l_ref, u_ref, l_cuda, u_cuda);
+        auto u_transpose_lin_op_ref = (*u_ref)->transpose();
+        auto u_transpose_csr_ref =
+            static_unique_ptr_cast<Csr>(std::move(u_transpose_lin_op_ref));
+        auto u_transpose_lin_op_cuda = (*u_cuda)->transpose();
+        auto u_transpose_csr_cuda =
+            static_unique_ptr_cast<Csr>(std::move(u_transpose_lin_op_cuda));
+
+        gko::kernels::reference::par_ilu_factorization::compute_l_u_factors(
+            ref, iterations, gko::lend(coo_ref), gko::lend(*l_ref),
+            gko::lend(u_transpose_csr_ref));
+        gko::kernels::cuda::par_ilu_factorization::compute_l_u_factors(
+            cuda, iterations, gko::lend(coo_cuda), gko::lend(*l_cuda),
+            gko::lend(u_transpose_csr_cuda));
+        auto u_lin_op_ref = u_transpose_csr_ref->transpose();
+        *u_ref = static_unique_ptr_cast<Csr>(std::move(u_lin_op_ref));
+        auto u_lin_op_cuda = u_transpose_csr_cuda->transpose();
+        *u_cuda = static_unique_ptr_cast<Csr>(std::move(u_lin_op_cuda));
+    }
+};
+
+
+TEST_F(ParIlu, KernelInitializeRowPtrsLUEquivalentToRef)
+{
+    auto num_row_ptrs = csr_ref->get_size()[0] + 1;
+    gko::Array<index_type> l_row_ptrs_array_ref(ref, num_row_ptrs);
+    gko::Array<index_type> u_row_ptrs_array_ref(ref, num_row_ptrs);
+    gko::Array<index_type> l_row_ptrs_array_cuda(cuda, num_row_ptrs);
+    gko::Array<index_type> u_row_ptrs_array_cuda(cuda, num_row_ptrs);
+
+    initialize_row_ptrs(
+        l_row_ptrs_array_ref.get_data(), u_row_ptrs_array_ref.get_data(),
+        l_row_ptrs_array_cuda.get_data(), u_row_ptrs_array_cuda.get_data());
+
+    GKO_ASSERT_ARRAY_EQ(&l_row_ptrs_array_ref, &l_row_ptrs_array_cuda);
+    GKO_ASSERT_ARRAY_EQ(&u_row_ptrs_array_ref, &u_row_ptrs_array_cuda);
+}
+
+
+TEST_F(ParIlu, KernelInitializeParILUIsEquivalentToRef)
+{
+    std::unique_ptr<Csr> l_ref{};
+    std::unique_ptr<Csr> u_ref{};
+    std::unique_ptr<Csr> l_cuda{};
+    std::unique_ptr<Csr> u_cuda{};
+
+    initialize_lu(&l_ref, &u_ref, &l_cuda, &u_cuda);
+
+    GKO_ASSERT_MTX_NEAR(l_ref, l_cuda, 1e-14);
+    GKO_ASSERT_MTX_NEAR(u_ref, u_cuda, 1e-14);
+}
+
+
+TEST_F(ParIlu, KernelComputeParILUIsEquivalentToRef)
+{
+    std::unique_ptr<Csr> l_ref{};
+    std::unique_ptr<Csr> u_ref{};
+    std::unique_ptr<Csr> l_cuda{};
+    std::unique_ptr<Csr> u_cuda{};
+
+    compute_lu(&l_ref, &u_ref, &l_cuda, &u_cuda);
+
+    GKO_ASSERT_MTX_NEAR(l_ref, l_cuda, 5e-2);
+    GKO_ASSERT_MTX_NEAR(u_ref, u_cuda, 5e-2);
+}
+
+
+TEST_F(ParIlu, KernelComputeParILUWithMoreIterationsIsEquivalentToRef)
+{
+    std::unique_ptr<Csr> l_ref{};
+    std::unique_ptr<Csr> u_ref{};
+    std::unique_ptr<Csr> l_cuda{};
+    std::unique_ptr<Csr> u_cuda{};
+    gko::size_type iterations{200};
+
+    compute_lu(&l_ref, &u_ref, &l_cuda, &u_cuda, iterations);
+
+    GKO_ASSERT_MTX_NEAR(l_ref, l_cuda, 1e-14);
```

Comment:
This likely won't be achieved.

Reply:
I tested it, and it does because it is doing 200 iterations.